### PR TITLE
npm support

### PR DESCRIPTION
**Pull request by asafgabai**
- [x] All [tests](https://github.com/jfrog/build-info-go#tests) passed. If this feature is not already covered by the tests, I added new tests.
- [x] This pull request is on the dev branch.
- [x] I used gofmt for formatting the code before submitting the pull request.
-----

### CODE REVIEW

File: utils/npmutils.go [hunk 1]
```diff
@@ -0,0 +1,303 @@
...
+	select {
+	case err = <-errorChan:
+	default:
+		return
+	}
+
+	return
+}
+
+func createHandlerFunc(dep *entities.Dependency, dependenciesChan chan *entities.Dependency, traverseDependenciesFunc func(dependency *entities.Dependency) (bool, error)) func(threadId int) error {
```
Comment (eyalbe4) [inline utils/npmutils.go:92]: Let's add a description for this function.

File: utils/npmutils.go [hunk 2]
```diff
@@ -0,0 +1,303 @@
...
+type TypeRestriction int
+
+const (
+	DefaultRestriction TypeRestriction = iota
+	All
+	DevOnly
+	ProdOnly
+)
+
+func CalculateDependenciesList(typeRestriction TypeRestriction, executablePath, srcPath, moduleId string, npmArgs []string, traverseDependenciesFunc func(dependency *entities.Dependency) (bool, error), threads int, log Log) (dependenciesList []entities.Dependency, err error) {
```
Comment (eyalbe4) [inline utils/npmutils.go:28]: Let's add a description for this function.

File: utils/npmutils.go [hunk 3]
```diff
@@ -0,0 +1,303 @@
...
+}
+
+// Parses npm dependencies recursively and adds the collected dependencies to the given dependencies map.
+func parseDependencies(data []byte, scope string, pathToRoot []string, dependencies *map[string]*entities.Dependency, log Log) error {
+	return jsonparser.ObjectEach(data, func(key []byte, value []byte, dataType jsonparser.ValueType, offset int) error {
+		depName := string(key)
+		ver, _, _, err := jsonparser.Get(data, depName, "version")
+		depVersion := string(ver)
+		depId := depName + ":" + depVersion
+		if err != nil && err != jsonparser.KeyPathNotFoundError {
```
Comment (eyalbe4) [inline utils/npmutils.go:140]: It is best to check the error right after (one line after) it is received.

File: utils/npmutils.go [hunk 4]
```diff
@@ -0,0 +1,303 @@
...
+
+func ReadPackageInfoFromPackageJson(packageJsonDirectory string, npmVersion *version.Version) (*PackageInfo, error) {
+	packageJson, err := ioutil.ReadFile(filepath.Join(packageJsonDirectory, "package.json"))
+	if err != nil {
+		return nil, err
+	}
+	return ReadPackageInfo(packageJson, npmVersion)
+}
+
+func ReadPackageInfo(data []byte, npmVersion *version.Version) (*PackageInfo, error) {
```
Comment (eyalbe4) [inline utils/npmutils.go:255]: Looks like thi sfunction can be made provate.

File: README.md [hunk 5]
```diff
@@ -68,22 +68,33 @@ All you need to do is to navigate to the project's root directory and run one of
...
 bi gradle
 ```
 
+#### npm
+
+```shell
+bi npm
+```
+
+Note: checksums calculation is not supported for npm projects.
```
Comment (eyalbe4) [inline README.md:90]: ...is not supported... --> ...is not yet supported...

**Review by eyalbe4 — CHANGES_REQUESTED** [review]

**Review by eyalbe4 — CHANGES_REQUESTED** [review]
Looks loke the code base includes the following line - 
```
fmt.Println("sssssssssssss")
```
Let's remove it.

File: utils/npmutils.go (hunk 4 above)
Comment (asafgabai) [reply utils/npmutils.go:255]: It's also used jfrog-cli-core.

**Pull request merged**